File: Binary Tree - DFS/Medium/1372. Longest ZigZag Path in a Binary Tree/longest_zigzag_path_in_a_binary_tree.py
# DFS
# T.C.-> O(n)
# S.C.-> O(n)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        self.maxLength = 0  # instance variable initialized here

        self.zigzagSearch(root, True, 0)
        self.zigzagSearch(root, False, 0)

        return self.maxLength

    def zigzagSearch(self, current, moveLeft, length):
        if not current:
            return

        self.maxLength = max(self.maxLength, length)

        if moveLeft:
            self.zigzagSearch(current.left, False, length + 1)
            self.zigzagSearch(current.right, True, 1)
        else:
            self.zigzagSearch(current.left, False, 1)
            self.zigzagSearch(current.right, True, length + 1)

File: Binary Tree - DFS/Medium/1372. Longest ZigZag Path in a Binary Tree/longest_zigzag_path_in_a_binary_tree.cpp
// DFS
// T.C.-> O(n)
// S.C.-> O(n)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxLength = 0;
    void zigzagSearch(TreeNode* current, bool moveLeft, int length) {
        if (current == nullptr) {
            return;
        }
        maxLength = max(maxLength, length);
        if (moveLeft) {
            zigzagSearch(current->left, false, length + 1);
            zigzagSearch(current->right, true, 1);
        } else {
            zigzagSearch(current->left, false, 1);
            zigzagSearch(current->right, true, length + 1);
        }
    }
    int longestZigZag(TreeNode* root) {
        zigzagSearch(root, true, 0);
        return maxLength;
    }
};
